File: bindings/nodejs/examples/password-hashing.js
'use strict';

// crypto_pwhash (Argon2id): hash and verify a password.
//
// PWHASH_INTERACTIVE is used here so the example runs fast - PWHASH_MODERATE (the default
// strength most applications should use) and PWHASH_SENSITIVE both take real seconds by design.
//
// Run: node examples/password-hashing.js

const assert = require('node:assert/strict');
const dstu = require('../js/index.js');

function main() {
  const password = Buffer.from('correct horse battery staple');
  const stored = dstu.pwhashHashPassword(password, dstu.PWHASH_INTERACTIVE);
  console.log(`stored hash: ${stored}`);

  assert.strictEqual(dstu.pwhashVerifyPassword(password, stored), true);
  console.log('correct password accepted');

  assert.strictEqual(dstu.pwhashVerifyPassword(Buffer.from('wrong guess'), stored), false);
  console.log('wrong password correctly rejected');
}

main();

File: bindings/go/examples/password_hashing.go
package main

import (
	"fmt"

	"github.com/user137/uacrypt/bindings/go/dstu"
)

// crypto_pwhash (Argon2id): hash and verify a password.
//
// PwhashInteractive is used here so the example runs fast - PwhashModerate (the strength most
// applications should use) and PwhashSensitive both take real seconds by design.
func runPasswordHashing() error {
	password := []byte("correct horse battery staple")
	stored, err := dstu.HashPassword(password, dstu.PwhashInteractive)
	if err != nil {
		return err
	}
	fmt.Printf("stored hash: %s\n", stored)

	if !dstu.VerifyPassword(password, stored) {
		return fmt.Errorf("correct password was rejected")
	}
	fmt.Println("correct password accepted")

	if !dstu.VerifyPassword([]byte("wrong guess"), stored) {
		fmt.Println("wrong password correctly rejected")
	}
	return nil
}
